### PR TITLE
[CU-332pb80] ensure slashed amount is greater than existential deposit

### PR DESCRIPTION
**Pull request by andresmechali**
## Issue
https://app.clickup.com/t/332pb80

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-332pb80 Slashed amount of minimum reward must be >= existential deposit](https://app.clickup.com/t/332pb80)

File: code/parachain/frame/staking-rewards/src/lib.rs [hunk 1]
```diff
@@ -248,6 +249,8 @@ pub mod pallet {
 		// NOTE(benluelo): This should be removed once this issue gets resolved:
 		// https://github.com/paritytech/substrate/issues/12257
 		NoDurationPresetsProvided,
+		/// Slashed amount of minimum reward is smaller than existential deposit
```
Comment (benluelo) [inline code/parachain/frame/staking-rewards/src/lib.rs:252]:
```suggestion
		/// Slashed amount of minimum reward is less than than existential deposit
```

File: code/parachain/frame/staking-rewards/src/lib.rs [hunk 2]
```diff
@@ -697,13 +702,24 @@ pub mod pallet {
 
 					// TODO: Replace into_iter with iter_mut once it's available
 					let rewards = initial_reward_config
+						.clone()
```
Comment (benluelo) [inline code/parachain/frame/staking-rewards/src/lib.rs:705]: This clone can be removed, move the `ensure!` above this statement

**Comment by benluelo** [comment]
A failure test should be added as well

File: code/parachain/frame/staking-rewards/src/lib.rs [hunk 3]
```diff
@@ -76,6 +76,7 @@ pub mod pallet {
 		},
 		time::{DurationSeconds, ONE_MONTH, ONE_WEEK},
 	};
+	use core::ops::Mul;
```
Comment (dzmitry-lahoda) [inline code/parachain/frame/staking-rewards/src/lib.rs:79]: sp_std instead of core?

**Review by dzmitry-lahoda — APPROVED** [review]

File: code/parachain/frame/staking-rewards/src/lib.rs [hunk 4]
```diff
@@ -695,6 +699,24 @@ pub mod pallet {
 
 					let now_seconds = T::UnixTime::now().as_secs();
 
+					ensure!(
+						initial_reward_config.iter().all(|(asset_id, reward_config)| {
+							if reward_config.reward_rate.amount > T::Balance::zero() {
+								// If none zero reward, check that it's greater than ED
```
Comment (benluelo) [inline code/parachain/frame/staking-rewards/src/lib.rs:705]:
```suggestion
								// If non-zero reward, check that the slashed amount is greater than ED
```

File: code/parachain/frame/staking-rewards/src/lib.rs [hunk 5]
```diff
@@ -695,6 +699,24 @@ pub mod pallet {
...
+								lock.unlock_penalty
+									.left_from_one()
+									.mul(reward_config.reward_rate.amount) >=
+									T::ExistentialDeposits::get(asset_id)
+							} else {
+								// Else, return true so check passes
+								// NOTE(connor): This is a band-aid that some better type management
+								// would remove the need for, outside the scope of this PR
```
Comment (benluelo) [inline code/parachain/frame/staking-rewards/src/lib.rs:713]: clickup task for this issue: https://app.clickup.com/t/2yz3j4r

**Review by PoisonPhang — APPROVED** [review]

**Review by benluelo — APPROVED** [review]

**Comment by benluelo** [comment]
mergify refresh

**Pull request merged**